client/src/controllers/PreviewController.ts

```diff
@@ -0,0 +1,789 @@
+) => {
+  const a11yRowTemplate = document.querySelector<HTMLTemplateElement>(
+    '#w-a11y-result-row-template',
+  );
+  const checksPanel = document.querySelector<HTMLElement>(
+    '[data-checks-panel]',
+  );
+  const config = getAxeConfiguration(document.body);
+  const toggleCounter = document.querySelector<HTMLElement>(
+    '[data-side-panel-toggle="checks"] [data-side-panel-toggle-counter]',
+  );
+  const panelCounter = document.querySelector<HTMLElement>(
+    '[data-side-panel="checks"] [data-a11y-result-count]',
+  );
+
+  if (
+    !checksPanel ||
+    !a11yRowTemplate ||
+    !config ||
+    !toggleCounter ||
+    !panelCounter
+  ) {
+    return;
+  }
+
+  // Ensure we only test within the preview iframe, but nonetheless with the correct selectors.
+  config.context = {
+    include: {
+      fromFrames: ['#w-preview-iframe'].concat(
+        (config.context as ContextObject).include as string[],
+      ),
+    },
+  } as ContextObject;
+  if ((config.context.exclude as string[])?.length > 0) {
+    config.context.exclude = {
+      fromFrames: ['#w-preview-iframe'].concat(
+        config.context.exclude as string[],
+      ),
+    } as ContextObject['exclude'];
+  }
+
+  const { results, a11yErrorsNumber } = await getA11yReport(config);
+
+  toggleCounter.innerText = a11yErrorsNumber.toString();
+  toggleCounter.hidden = a11yErrorsNumber === 0;
+  panelCounter.innerText = a11yErrorsNumber.toString();
+  panelCounter.classList.toggle('has-errors', a11yErrorsNumber > 0);
+
+  renderA11yResults(
+    checksPanel,
+    results,
+    config,
+    a11yRowTemplate,
+    onClickSelector,
+  );
+};
+
+interface PreviewDataResponse {
+  is_valid: boolean;
+  is_available: boolean;
+}
+
+/**
+ * Controls the preview panel component to submit the current form state and
+ * update the preview iframe if the form is valid.
+ *
+ * Dispatches the following events, in order:
+ * - `update` before sending the preview data to the server. Cancelable.
+ * - `json` after the preview data update request is completed. Note that the
+ *   data may not be valid and the iframe is not reloaded yet at this point.
+ *   Event `detail`:
+ *   - `data`: the response data that indicates whether the submitted data was
+ *     valid and whether the preview is available.
+ * - `error` when an error occurs while updating the preview data.
+ *   Event `detail`:
+ *   - `error`: the error object that was thrown.
+ * - `load` before reloading the preview iframe. Cancelable.
+ * - `loaded` after the preview iframe has been reloaded.
+ * - `ready` when the preview is ready for further updates – only fired on initial load.
+ * - `updated` after an update cycle is finished. This may or may not involve a
+ *   refresh of the iframe (e.g. in the case of an invalid form).
+ */
+export class PreviewController extends Controller<HTMLElement> {
+  static classes = ['hasErrors', 'selectedSize'];
+
+  static targets = ['size', 'newTab', 'spinner', 'mode', 'iframe'];
```

Comment:
```suggestion
  static targets = ['iframe', 'mode', 'newTab', 'size', 'spinner'];
```

Suggestion for alpha ordering here.